tests/30rooms/06invite.pl

```diff
@@ -107,3 +111,424 @@
          Future->done(1);
       });
    };
+
+test "Can invite existing 3pid",
+   requires => [qw( user more_users test_http_server_hostandport await_http_request )],
+
+   do => sub {
+      my ( $inviter, $more_users, $id_server, $await_http_request ) = @_;
+
+      my $invitee_email = "marmosets\@monkeyworld.org";
+      my $invitee_mxid = $more_users->[0]->user_id;
+      my $room_id;
+
+      Future->needs_all(
+         stub_is_lookup($invitee_email, $invitee_mxid, $await_http_request),
+
+         matrix_create_and_join_room( [ $inviter ], visibility => "private" )
+         ->then( sub {
+            ( $room_id ) = @_;
+            do_request_json_for( $inviter,
+               method => "POST",
+               uri    => "/api/v1/rooms/$room_id/invite",
+
+               content => {
+                  id_server => $id_server,
+                  medium => "email",
+                  address => $invitee_email,
+                  display_name => "Cute things",
+               },
+            ),
+         })->then( sub {
+            matrix_get_room_state( $inviter, $room_id,
+               type => "m.room.member",
+               state_key => $invitee_mxid,
+            )->on_done( sub {
+               my ( $body ) = @_;
+
+               log_if_fail "Body", $body;
+               $body->{membership} eq "invite" or
+                  die "Expected invited user membership to be 'invite'";
+            });
+         }),
+      );
+   };
+
+test "Can invite unbound 3pid",
+   requires => [qw( user more_users test_http_server_hostandport await_http_request first_home_server synapse_client_locations )],
+   do => sub {
+      my ( $user, $users, $id_server, $await_http_request, $user_agent ) = @_;
```

Comment:
You appear to be 'requires'ing more things than you actually unpack out of `@_` here.